README.md
# FOCUSTOOLS
Python utilities supporting the [FOCUS package](http://focus-em.org) and cryo-EM data processing tasks.

__FOCUSTOOLS__ offers Python functions and scripts for performing common tasks in cryogenic electron microscopy (cryo-EM) data processing such as postprocessing 3D maps, filtering, masking, computing FSC curves, cropping/padding in real and Fourier space, CTF correction, among others. Some of the scripts and functions contained here were developed to support the FOCUS package. Others were written for the developer's own studies or mere convenience, and are made available here in the hope they can be useful to someone else.

The I/O of MRC files (with or without compression) is based on the [MRCZ library](https://github.com/em-MRCZ/python-mrcz).

Where possible, tasks are accelerated by the [NumExpr](https://github.com/pydata/numexpr) engine.

## Installation
__FOCUSTOOLS__ can be installed from [PyPI](http://pypi.org) using `pip`:

`pip install focustools`

If not using `sudo` (like above), it will be installed locally in the user `HOME`. In this case, make sure you have the following location in your `PATH` environment variable in order to be able to call the scripts directly:
`~/.local/bin/`

## Contents

Please see the [release notes](RELEASE_NOTES.md).

## Usage

Once installed, the scripts can be called directly from your shell, for example:

`focus.postprocess half1.mrc half2.mrc --angpix 0.639 --automask --auto_bfac -1,-1 --mtf data_mtf_k2_300kv.star --out postprocessed`

Use the `--help` flag to obtain a comprehensive list of options to the desired script, e.g.

`focus.postprocess --help`

### Tutorials

For more details, see our [Jupyter notebook](http://jupyter.org/install) tutorials:

* [`focus.postprocess`](tutorial/tutorial_postprocess.ipynb) script
* [`focus.mrcz`](tutorial/tutorial_mrcz.ipynb) script
* [`focus.ctf`](tutorial/tutorial_ctf.ipynb) script
* [`focustools`](tutorial/tutorial_focustools.ipynb) module: a brief guide to using the Python functions implemented in __FOCUSTOOLS__

## Citing

If __FOCUSTOOLS__ in general is useful in your work, please cite:
* Biyani, N., Righetto, R.D., McLeod, R., Caujolle-Bert, D., Castano-Diez, D., Goldie, K.N., Stahlberg, H., 2017. _Focus: The interface between data collection and data processing in cryo-EM_. J. Struct. Biol. 198, 124–133. https://doi.org/10.1016/j.jsb.2017.03.007

If `focus.postprocess` is useful in your work, please cite:
* Righetto, R.D., Biyani, N., Kowal, J., Chami, M., Stahlberg, H., 2019. _Retrieving high-resolution information from disordered 2D crystals by single-particle cryo-EM_. Nat. Commun. 10, 1722. https://doi.org/10.1038/s41467-019-09661-5

If `focus.mrcz` or the [MRCZ library](https://github.com/em-MRCZ/python-mrcz) are useful in your work, please cite:
* McLeod, R.A., Diogo Righetto, R., Stewart, A., Stahlberg, H., 2018. _MRCZ – A file format for cryo-TEM data with fast compression_. J. Struct. Biol. 201, 252–257. https://doi.org/10.1016/j.jsb.2017.11.012

References for specific methods implemented in __FOCUSTOOLS__ that are published elsewhere can be obtained by calling the scripts with the `--help` option or from the source code directly.

## Reporting bugs & Troubleshooting

Please use the [issue tracker](https://github.com/C-CINA/focustools/issues) to report bugs or ask for help, suggest features, etc.

## Authors
__FOCUSTOOLS__ was initially developed by [Ricardo Righetto](https://github.com/rdrighetto/) at the [Stahlberg lab](http://c-cina.org).  
The `FCC()` function was written by Robert A. McLeod.



RELEASE_NOTES.md
# 0.2.3

* Fixed error in CTF formula of `CTF()` of `focustools/util.py` - thanks to Harshit Gupta for spotting it

# 0.2.2

* Fixed bug in `focus.postprocess` (reading masks)

# 0.2.1

* Fixed bug in `focus.postprocess` (reading half-maps)

# 0.2.0

* Renamed `bin/focus.ctfapply` to `bin/focus.ctf`
* Added option `--randomize_beyond` to `focus.postprocess`
* Added options `--transparent` and `--svg` to `focus.ctf`
* Included `bin/focus.resample`
* Included `bin/focus.project`

# 0.1.4

Fixed default argument bugs in `AutoMask()` of `focustools/util.py`

# 0.1.3

Fixed x-label in CTF plots of `bin/focus.ctfapply`

# 0.1.2

Another bugfix in `focustools/util.py`

# 0.1.1

Bug fixes in:

* `bin/focus.ctfapply`
* `bin/focus.mrcz`
* `focustools/util.py`
* `RELEASE_NOTES.md`

# 0.1.0

Initial release containing:

* `bin/`  
  * focus.ctfapply  
  * focus.mrcz  
  * focus.postprocess  

* `focustools/utils.py`  
  * AdhocSSNR(imsize=[100, 100], apix=1.0, DF=1000.0, WGH=0.1, Cs=2.7, kV=300.0, S=1.0, F=1.0, hp_frac=0.01, lp=True)  
  * AutoMask(img, apix=1.0, lp=-1, gaussian=False, cosine=True, cosine_edge_width=3.0, absolute_threshold=None, fraction_threshold=None, sigma_threshold=1.0, expand_width=3.0, expand_soft_width=3.0, floodfill_rad=-1, floodfill_xyz=[0, 0, 0], floodfill_fraction=-1, verbose=False)  
  * BackProject(img, pose=[0, 0, 0, 0, 0], interpolation='trilinear', pad=2, do_sinc=True, res_max=-1, apix=-1, return_weights=True, is_fft=False, DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, invert_contrast=False, Cs=2.7, kV=300.0, phase_flip=False, ctf_multiply=False, wiener_filter=False, C=1.0)  
  * BandPassCrossCorrelation(img1, img2, apix=1.0, lp=-1, hp=-1, weights=None, rounding=True, eps=1e-8)  
  * CistemAutoMask(img, apix=1.0)  
  * CorrectCTF(img, DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, invert_contrast=False, Cs=2.7, kV=300.0, apix=1.0, phase_flip=False, ctf_multiply=False, wiener_filter=False, C=1.0, return_ctf=False)  
  * CrossCorrelation(img1, img2)  
  * CTF(imsize=[100, 100], DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, Cs=2.7, kV=300.0, apix=1.0, B=0.0, rfft=True)  
  * ElectronWavelength(kV=300.0)  
  * FCC(volume1, volume2, phiArray=[0.0], invertCone=False)  
  * FilterBfactor(img, apix=1.0, B=0.0, return_filter=False)  
  * FilterCosine(img, apix=1.0, lp=-1, hp=-1, width=6.0, return_filter=False)  
  * FilterDoseWeight(stack, apix=1.0, frame_dose=1.0, pre_dose=0.0, total_dose=-1, kv=300.0)  
  * FilterGauss(img, apix=1.0, lp=-1, hp=-1, return_filter=False)  
  * FilterTophat(img, apix=1.0, lp=-1, hp=-1, return_filter=False)  
  * FilterWhiten(img, return_filter=False, ps=False)  
  * FirstZeroCTF(DF=1000.0, WGH=0.10, Cs=2.7, kV=300.0)  
  * FloodFilling(img, inimask, thr=0.0)  
  * FRC(image1, image2, phiArray=[0.0])  
  * FSC(volume1, volume2, phiArray=[0.0])  
  * Fsc2Xml(filename, x, y)  
  * GetFreqArray(NSAM=256, apix=1.0)  
  * GetNumberOfFourierSamples(imsize)  
  * HalfBitCurve(imsize=[100, 100], nsym=1, D=2.0, L=3.0, count=False)  
  * HighResolutionNoiseSubstitution(img, apix=1.0, lp=-1, random_seed=123456789, parallel=False)  
  * MaskAutoExpand(map1, map2, apix=1.0, init_lp=15, fraction_threshold=0.01, init_hard_edge=0, init_soft_edge=3, randomize_below_fsc=0.8, fsc_thr=0.143, random_seed=123456789, interp=False, verbose=False)  
  * MatchAmplitudes(img1, img2)  
  * MatchPowerSpectra(img1, img2)  
  * NormalizeImg(img, mean=0.0, std=1.0, radius=-1)  
  * Project(img, pose=[0, 0, 0, 0, 0], interpolation='trilinear', pad=2, do_sinc=True, res_max=-1, apix=-1, is_fft=False, DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, invert_contrast=False, Cs=2.7, kV=300.0, phase_flip=False, ctf_multiply=False)  
  * RadialFilter(img, filt, return_filter=False)  
  * RadialIndices(imsize=[100, 100], rounding=True, normalize=False, rfft=False, xyz=[0, 0, 0], nozero=True)  
  * RadialProfile(img, amps=False)  
  * Resample(img, newsize=None, apix=1.0, newapix=None)  
  * Resize(img, newsize=None, padval=None, xyz=[0, 0, 0])  
  * ResolutionAtThreshold(freq, fsc, thr, interp=True, nyquist_is_fine=False, )  
  * Rotate(img, rot=[0, 0, 0], interpolation='trilinear', pad=1)  
  * RotateFFT(img, rot=[0, 0, 0], interpolation='trilinear', pad=1)  
  * RotationalAverage(img, nomean=False)  
  * Shift(img, shift=[0, 0, 0])  
  * SigmaCurve(imsize=[100, 100], sigma=3.0, nsym=1, D=2.0, L=3.0, count=False)  
  * SoftMask(imsize=[100, 100], radius=0.5, width=6.0, rounding=False, xyz=[0, 0, 0], rfft=False)  
  * VoxelsPerShell(imsize=[100, 100], count=False)  


focustools/CTF.py
# Python utilities for CTF correction
# Author: Ricardo Righetto
# E-mail: ricardo.righetto@unibas.ch

import numpy as np
from focus_utilities import util
import copy
import numexpr as ne

pi = np.pi

def CTF(imsize=[100, 100], DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, Cs=2.7, kV=300.0, apix=1.0, B=0.0, rfft=True):
    # Generates 2D CTF function
    # Underfocus is positive following conventions of FREALIGN and most of the packages out there (in Angstroms).
    # B is B-factor
    # rfft is to compute only half of the FFT (i.e. real data) if True, or the full FFT if False.

    # try:

    #     dummy = np.fft.rfftfreq.func_name

    # except AttributeError:

    #     raise AttributeError("""\nERROR: Your version of NumPy does not contain numpy.fft.rfftfreq. Please switch to NumPy version 1.8.0 or later.\nSometimes this error occurs due to the Python environment being overshadowed by another program such as EMAN2, for example.\nIf you are using FOCUS from the GUI you can check this under Settings >> Software.""")

    if not np.isscalar(imsize) and len(imsize) == 1:

        imsize = imsize[0]

    Cs *= 1e7  # Convert Cs to Angstroms

    if DF2 == None or np.isscalar(imsize):

        DF2 = DF1

    # else:

    #   # NOTATION FOR DEFOCUS1, DEFOCUS2, ASTIGMASTISM BELOW IS INVERTED DUE TO NUMPY CONVENTION:
    #   DF1, DF2 = DF2, DF1

    AST *= -pi / 180.0

    WL = ElectronWavelength(kV)

    w1 = np.sqrt(1 - WGH * WGH)
    w2 = WGH

    import warnings
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=RuntimeWarning)

        if np.isscalar(imsize):

            if rfft:
                rmesh = np.fft.rfftfreq(imsize)
            else:
                rmesh = np.fft.fftfreq(imsize)
            amesh = 0.0

        else:

            # rmesh,amesh = focus_utilities.RadialIndices( imsize, RFFT=True )
            xmesh = np.fft.fftfreq(imsize[0])
            if rfft:
                ymesh = np.fft.rfftfreq(imsize[1])
            else:
                ymesh = np.fft.fftfreq(imsize[1])

            xmeshtile = np.tile(xmesh, [len(ymesh), 1]).T
            ymeshtile = np.tile(ymesh, [len(xmesh), 1])

            # rmesh = np.sqrt(xmeshtile * xmeshtile +
            #                 ymeshtile * ymeshtile) / apix
            rmesh = ne.evaluate("sqrt(xmeshtile * xmeshtile + ymeshtile * ymeshtile) / apix")

            amesh = np.nan_to_num(ne.evaluate("arctan2(ymeshtile, xmeshtile)"))

        rmesh2 = ne.evaluate("rmesh * rmesh")

        # From Mindell & Grigorieff, JSB 2003:
        DF = ne.evaluate("0.5 * (DF1 + DF2 + (DF1 - DF2) * cos(2.0 * (amesh - AST)))")

        Xr = np.nan_to_num(ne.evaluate("pi * WL * rmesh2 * (DF - 1.0 / (2.0 * WL * WL * rmesh2 * Cs))"))

    # sinXr = ne.evaluate("sin(Xr)")
    # cosXr = ne.evaluate("cos(Xr)")
    # CTFreal = w1 * sinXr - w2 * cosXr
    # CTFimag = -w1 * cosXr - w2 * sinXr

    # CTFim = CTFreal + CTFimag*1j
    CTFim = ne.evaluate("-w1 * sin(Xr) - w2 * cos(Xr)")

    if B != 0.0:  # Apply B-factor only if necessary:

        ne.evaluate("CTFim * np.exp(-B * (rmesh2) / 4)", out=CTFim)

    return CTFim


def ElectronWavelength(kV=300.0):
    # Returns electorn wavelength in Angstroms
    # kV *= 1e3 # ensure Kilovolts for below formula
    # return 12.2639 / np.sqrt( kV + 0.97845 * kV*kV / ( 1e6 ) )
    return 12.2639 / np.sqrt(kV * 1e3 + 0.97845 * kV * kV)


def FirstZeroCTF(DF=1000.0, WGH=0.10, Cs=2.7, kV=300.0):
    # Finds the resolution at the first zero of the CTF
    # Wolfram Alpha, solving for -w1 * sinXr - w2 * cosXr = 0
    # https://www.wolframalpha.com/input/?i=solve+%CF%80*L*(g%5E2)*(d-1%2F(2*(L%5E2)*(g%5E2)*C))%3Dn+%CF%80+-+tan%5E(-1)(c%2Fa)+for+g

    Cs *= 1e7  # Convert Cs to Angstroms

    w1 = np.sqrt(1 - WGH * WGH)
    w2 = WGH

    WL = ElectronWavelength(kV)

    g = np.sqrt(-2 * Cs * WL * np.arctan2(w2, w1) + 2 * pi *
                Cs * WL + pi) / (np.sqrt(2 * pi * Cs * DF) * WL)

    return g


def CorrectCTF(img, DF1=1000.0, DF2=None, AST=0.0, WGH=0.10, invert_contrast=False, Cs=2.7, kV=300.0, apix=1.0, phase_flip=False, ctf_multiply=False, wiener_filter=False, C=1.0, return_ctf=False):
    # Applies CTF correction to image
    # Type can be one of the following:
    # 0 - Phase-flipping only
    # 1 - CTF multiplication
    # 2 - Wiener filtering with Wiener constant C
    # By default will return images with same contrast as input, otherwise set invert_contrast=True.

    # Direct CTF correction would invert the image contrast. By default we don't do that, hence the negative sign:
    CTFim = -CTF(img.shape, DF1, DF2, AST, WGH, Cs, kV, apix, 0.0, rfft=True)

    CTFcor = []
    cortype = []

    if invert_contrast:

        ne.evaluate("CTFim * -1.0", out=CTFim)

    FT = np.fft.rfftn(img)

    if phase_flip:  # Phase-flipping

        s = np.sign(CTFim)
        CTFcor.append(np.fft.irfftn(ne.evaluate("FT * s")))
        cortype.append('pf')

    if ctf_multiply:  # CTF multiplication

        CTFcor.append(np.fft.irfftn(ne.evaluate("FT * CTFim")))
        cortype.append('cm')

    if wiener_filter:  # Wiener filtering

        if np.any(C <= 0.0):

            raise ValueError(
                "Error: Wiener filter constant cannot be less than or equal to zero! C = %f " % C)

        CTFcor.append(np.fft.irfftn(ne.evaluate("FT * CTFim / (CTFim * CTFim + C)")))
        cortype.append('wf')

    # else:

    #   raise ValueError( "Error: Type of CTF correction must be 0 (phase-flipping), 1 (CTF multiplication) or 2 (Wiener filtering). ctftype = %d " % ctftype )

    # if return_half:

    #   # AmpHalf = np.zeros( img.shape )
    #   CTFnorm = CTFim**2
    #   AmpHalf = np.abs( FT )**2
    #   AmpHalf = AmpHalf - AmpHalf.mean() + CTFnorm.mean()
    #   AmpHalf = AmpHalf*CTFnorm.std()/AmpHalf.std()
    #   # CTFnorm -= CTFnorm.mean()
    #   # CTFnorm /= CTFnorm.std()
    #   AmpHalf[:,img.shape[1]/2:] = CTFnorm[:,img.shape[1]/2:]
    #   # AmpHalf = AmpHalf**2

    # if return_ctf and return_half:

    #   return CTFcor.real, CTFim, AmpHalf

    # elif return_ctf:

    #   return CTFcor.real, CTFim

    # elif return_half:

    #   return CTFcor.real, AmpHalf

    if return_ctf:

        CTFcor.append(CTFim)

    CTFcor.append(cortype)

    return CTFcor


def AdhocSSNR(imsize=[100, 100], apix=1.0, DF=1000.0, WGH=0.1, Cs=2.7, kV=300.0, S=1.0, F=1.0, hp_frac=0.01, lp=True):
        # Ad hoc SSNR model for micrograph deconvolution as proposed by Dimitry Tegunov. For details see:
        # https://www.biorxiv.org/content/10.1101/338558v1
        # https://github.com/dtegunov/tom_deconv/blob/master/tom_deconv.m

    rmesh = util.RadialIndices(
        imsize, rounding=False, normalize=True, rfft=True)[0]
    ne.evaluate("rmesh / apix", out=rmesh)
    # The ad hoc SSNR exponential falloff
    falloff = ne.evaluate("exp(-100 * rmesh * F) * 10**(3 * S)")

    # The cosine-shaped high-pass filter. It starts at zero frequency and reaches 1.0 at hp_freq (fraction of the Nyquist frequency)
    a = np.minimum(1.0, ne.evaluate("rmesh * apix / hp_frac"))
    highpass = ne.evaluate("1.0 - cos(a * pi/2)")

    if lp:

        # Ensure the filter will reach zero at the first zero of the CTF
        first_zero_res = FirstZeroCTF(DF=DF, WGH=WGH, Cs=Cs, kV=kV)
        a = np.minimum(1.0, ne.evaluate("rmesh / first_zero_res"))
        lowpass = ne.evaluate("cos(a * pi/2)")

        ssnr = ne.evaluate("highpass * falloff * lowpass")  # Composite filter

    else:

        ssnr = ne.evaluate("highpass * falloff")  # Composite filter

    return ssnr
